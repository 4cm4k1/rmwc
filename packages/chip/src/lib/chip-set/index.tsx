import React, { useRef } from 'react';
import * as RMWC from '@rmwc/types';
import { createComponent, Tag, useClassNames } from '@rmwc/base';
import { useChipSetFoundation } from './foundation';
import { ChipContext, ChipContextT } from '../chip-context';

/*********************************************************************
 * Chip Set
 *********************************************************************/

/** A container for multiple chips. */
export interface ChipSetProps {
  /** Causes the chis to overflow instead of wrap (their default behavior). */
  overflow?: boolean;
  /** Creates a action chipset. */
  action?: boolean;
  /** Creates a input chipset. */
  input?: boolean;
  /** Creates a filter chipset. */
  filter?: boolean;
  /** Determines whether chipset should be multiple-select or single-select. This is only supported for filter chips. */
  multipleSelect?: boolean;
}

export type ChipSetHTMLProps = RMWC.HTMLProps<
  HTMLDivElement,
  Omit<React.AllHTMLAttributes<HTMLDivElement>, 'label'>
>;

/** A container for multiple chips. */
export const ChipSet: RMWC.ComponentType<
  ChipSetProps,
  ChipSetHTMLProps,
  'div'
> = createComponent<ChipSetProps, ChipSetHTMLProps>(
  function ChipSet(props, ref) {
    const { rootEl, registerChip, unregisterChip } =
      useChipSetFoundation(props);

    const {
      overflow,
      action = true,
      input,
      filter,
      multipleSelect,
      ...rest
    } = props;

    const className = useClassNames(props, [
      'mdc-evolution-chip-set',
      {
        'mdc-evolution-chip-set--overlow': overflow
      }
    ]);

    const otherProps = {
      'aria-orientation': 'horizontal',
      'aria-multiselectable': multipleSelect
    };

    const contextApi = useRef<ChipContextT>({
      registerChip,
      unregisterChip,
      action,
      input,
      filter
    });

    return (
      <ChipContext.Provider value={contextApi.current}>
        <Tag
          className={className}
          tag="span"
          role={filter ? 'listbox' : 'grid'}
          element={rootEl}
          {...(filter && otherProps)}
          {...rest}
          ref={ref}
        >
          <span className="mdc-evolution-chip-set__chips" role="presentation">
            {props.children}
          </span>
        </Tag>
      </ChipContext.Provider>
    );
  }
);
