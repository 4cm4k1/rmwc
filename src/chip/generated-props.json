{
  "npm": {},
  "typescript": {
    "ChipProps": {
      "documentation": {
        "contents": [
          "<p>A Chip component.</p>\n"
        ],
        "contentsRaw": "A Chip component.",
        "metadata": {}
      },
      "fileName": "src/chip/index.tsx",
      "flags": {
        "isExported": true,
        "isExternal": false,
        "isOptional": false,
        "isPrivate": false,
        "isProtected": false,
        "isPublic": false,
        "isRest": false,
        "isStatic": false
      },
      "kind": "interface",
      "name": "ChipProps",
      "methods": [],
      "properties": [
        {
          "documentation": {
            "contents": [
              "<p>Includes an optional checkmark for the chips selected state.</p>\n"
            ],
            "contentsRaw": "Includes an optional checkmark for the chips selected state.",
            "metadata": {}
          },
          "fileName": "src/chip/index.tsx",
          "flags": {
            "isExported": true,
            "isExternal": false,
            "isOptional": true,
            "isPrivate": false,
            "isProtected": false,
            "isPublic": false,
            "isRest": false,
            "isStatic": false
          },
          "kind": "property",
          "name": "checkmark",
          "type": "undefined | false | true"
        },
        {
          "documentation": {
            "contents": [
              "<p>Additional children will be rendered in the text area.</p>\n"
            ],
            "contentsRaw": "Additional children will be rendered in the text area.",
            "metadata": {}
          },
          "fileName": "src/chip/index.tsx",
          "flags": {
            "isExported": true,
            "isExternal": false,
            "isOptional": true,
            "isPrivate": false,
            "isProtected": false,
            "isPublic": false,
            "isRest": false,
            "isStatic": false
          },
          "kind": "property",
          "name": "children",
          "type": "React.ReactNode"
        },
        {
          "documentation": {
            "contents": [
              "<p>Advanced: A reference to the MDCFoundation.</p>\n"
            ],
            "contentsRaw": "Advanced: A reference to the MDCFoundation.",
            "metadata": {}
          },
          "fileName": "src/chip/index.tsx",
          "flags": {
            "isExported": true,
            "isExternal": false,
            "isOptional": true,
            "isPrivate": false,
            "isProtected": false,
            "isPublic": false,
            "isRest": false,
            "isStatic": false
          },
          "kind": "property",
          "name": "foundationRef",
          "type": "React.Ref<MDCChipFoundation>"
        },
        {
          "documentation": {
            "contents": [
              "<p>Instance of an Icon Component.</p>\n"
            ],
            "contentsRaw": "Instance of an Icon Component.",
            "metadata": {}
          },
          "fileName": "src/chip/index.tsx",
          "flags": {
            "isExported": true,
            "isExternal": false,
            "isOptional": true,
            "isPrivate": false,
            "isProtected": false,
            "isPublic": false,
            "isRest": false,
            "isStatic": false
          },
          "kind": "property",
          "name": "icon",
          "type": "RMWC.IconPropT"
        },
        {
          "documentation": {
            "contents": [
              "<p>An optional chip ID that will be included in callback evt.detail. If this is not passed, RMWC will attempt to use the &quot;key&quot; prop if present.</p>\n"
            ],
            "contentsRaw": "An optional chip ID that will be included in callback evt.detail. If this is not passed, RMWC will attempt to use the \"key\" prop if present.",
            "metadata": {}
          },
          "fileName": "src/chip/index.tsx",
          "flags": {
            "isExported": true,
            "isExternal": false,
            "isOptional": true,
            "isPrivate": false,
            "isProtected": false,
            "isPublic": false,
            "isRest": false,
            "isStatic": false
          },
          "kind": "property",
          "name": "id",
          "type": "undefined | string"
        },
        {
          "documentation": {
            "contents": [
              "<p>Text for your Chip.</p>\n"
            ],
            "contentsRaw": "Text for your Chip.",
            "metadata": {}
          },
          "fileName": "src/chip/index.tsx",
          "flags": {
            "isExported": true,
            "isExternal": false,
            "isOptional": true,
            "isPrivate": false,
            "isProtected": false,
            "isPublic": false,
            "isRest": false,
            "isStatic": false
          },
          "kind": "property",
          "name": "label",
          "type": "React.ReactNode"
        },
        {
          "documentation": {
            "contents": [
              "<p>A callback for click or enter key. This should be used over onClick for accessibility reasons. evt.detail = { chipId: string }</p>\n"
            ],
            "contentsRaw": "A callback for click or enter key. This should be used over onClick for accessibility reasons. evt.detail = { chipId: string }",
            "metadata": {}
          },
          "fileName": "src/chip/index.tsx",
          "flags": {
            "isExported": true,
            "isExternal": false,
            "isOptional": true,
            "isPrivate": false,
            "isProtected": false,
            "isPublic": false,
            "isRest": false,
            "isStatic": false
          },
          "kind": "property",
          "name": "onInteraction",
          "type": "undefined | (evt: ChipOnInteractionEventT) => void"
        },
        {
          "documentation": {
            "contents": [
              "<p>A callback that is fired once the chip is in an exited state from removing it. evt.detail = { chipId: string }</p>\n"
            ],
            "contentsRaw": "A callback that is fired once the chip is in an exited state from removing it. evt.detail = { chipId: string }",
            "metadata": {}
          },
          "fileName": "src/chip/index.tsx",
          "flags": {
            "isExported": true,
            "isExternal": false,
            "isOptional": true,
            "isPrivate": false,
            "isProtected": false,
            "isPublic": false,
            "isRest": false,
            "isStatic": false
          },
          "kind": "property",
          "name": "onRemove",
          "type": "undefined | (evt: ChipOnRemoveEventT) => void"
        },
        {
          "documentation": {
            "contents": [
              "<p>A callback for click or enter key for the trailing icon. material-components-web always treats this as an intent to remove the chip. evt.detail = { chipId: string }</p>\n"
            ],
            "contentsRaw": "A callback for click or enter key for the trailing icon. material-components-web always treats this as an intent to remove the chip. evt.detail = { chipId: string }",
            "metadata": {}
          },
          "fileName": "src/chip/index.tsx",
          "flags": {
            "isExported": true,
            "isExternal": false,
            "isOptional": true,
            "isPrivate": false,
            "isProtected": false,
            "isPublic": false,
            "isRest": false,
            "isStatic": false
          },
          "kind": "property",
          "name": "onTrailingIconInteraction",
          "type": "undefined | (evt: ChipOnTrailingIconInteractionEventT) => void"
        },
        {
          "documentation": {
            "contents": [
              "<p>makes the Chip appear selected.</p>\n"
            ],
            "contentsRaw": "makes the Chip appear selected.",
            "metadata": {}
          },
          "fileName": "src/chip/index.tsx",
          "flags": {
            "isExported": true,
            "isExternal": false,
            "isOptional": true,
            "isPrivate": false,
            "isProtected": false,
            "isPublic": false,
            "isRest": false,
            "isStatic": false
          },
          "kind": "property",
          "name": "selected",
          "type": "undefined | false | true"
        },
        {
          "documentation": {
            "contents": [
              "<p>Instance of an Icon Component.</p>\n"
            ],
            "contentsRaw": "Instance of an Icon Component.",
            "metadata": {}
          },
          "fileName": "src/chip/index.tsx",
          "flags": {
            "isExported": true,
            "isExternal": false,
            "isOptional": true,
            "isPrivate": false,
            "isProtected": false,
            "isPublic": false,
            "isRest": false,
            "isStatic": false
          },
          "kind": "property",
          "name": "trailingIcon",
          "type": "RMWC.IconPropT"
        },
        {
          "documentation": {
            "contents": [
              "<p>Defaults to true. Set this to false if your trailing icon is something other than a remove button.</p>\n"
            ],
            "contentsRaw": "Defaults to true. Set this to false if your trailing icon is something other than a remove button.",
            "metadata": {}
          },
          "fileName": "src/chip/index.tsx",
          "flags": {
            "isExported": true,
            "isExternal": false,
            "isOptional": true,
            "isPrivate": false,
            "isProtected": false,
            "isPublic": false,
            "isRest": false,
            "isStatic": false
          },
          "kind": "property",
          "name": "trailingIconRemovesChip",
          "type": "undefined | false | true"
        }
      ]
    },
    "ChipSetProps": {
      "documentation": {
        "contents": [
          "<p>A container for multiple chips.</p>\n"
        ],
        "contentsRaw": "A container for multiple chips.",
        "metadata": {}
      },
      "fileName": "src/chip/index.tsx",
      "flags": {
        "isExported": true,
        "isExternal": false,
        "isOptional": false,
        "isPrivate": false,
        "isProtected": false,
        "isPublic": false,
        "isRest": false,
        "isStatic": false
      },
      "kind": "interface",
      "name": "ChipSetProps",
      "methods": [],
      "properties": [
        {
          "documentation": {
            "contents": [
              "<p>Creates a choice chipset</p>\n"
            ],
            "contentsRaw": "Creates a choice chipset",
            "metadata": {}
          },
          "fileName": "src/chip/index.tsx",
          "flags": {
            "isExported": true,
            "isExternal": false,
            "isOptional": true,
            "isPrivate": false,
            "isProtected": false,
            "isPublic": false,
            "isRest": false,
            "isStatic": false
          },
          "kind": "property",
          "name": "choice",
          "type": "undefined | false | true"
        },
        {
          "documentation": {
            "contents": [
              "<p>Creates a filter chipset</p>\n"
            ],
            "contentsRaw": "Creates a filter chipset",
            "metadata": {}
          },
          "fileName": "src/chip/index.tsx",
          "flags": {
            "isExported": true,
            "isExternal": false,
            "isOptional": true,
            "isPrivate": false,
            "isProtected": false,
            "isPublic": false,
            "isRest": false,
            "isStatic": false
          },
          "kind": "property",
          "name": "filter",
          "type": "undefined | false | true"
        }
      ]
    },
    "ChipOnInteractionEventT": {
      "documentation": {
        "contents": [
          "<p>Events</p>\n"
        ],
        "contentsRaw": "Events",
        "metadata": {}
      },
      "fileName": "src/chip/index.tsx",
      "flags": {
        "isExported": true,
        "isExternal": false,
        "isOptional": false,
        "isPrivate": false,
        "isProtected": false,
        "isPublic": false,
        "isRest": false,
        "isStatic": false
      },
      "kind": "type alias",
      "name": "ChipOnInteractionEventT",
      "type": "RMWC.CustomEventT<>"
    },
    "ChipOnTrailingIconInteractionEventT": {
      "fileName": "src/chip/index.tsx",
      "flags": {
        "isExported": true,
        "isExternal": false,
        "isOptional": false,
        "isPrivate": false,
        "isProtected": false,
        "isPublic": false,
        "isRest": false,
        "isStatic": false
      },
      "kind": "type alias",
      "name": "ChipOnTrailingIconInteractionEventT",
      "type": "RMWC.CustomEventT<>"
    },
    "ChipOnRemoveEventT": {
      "fileName": "src/chip/index.tsx",
      "flags": {
        "isExported": true,
        "isExternal": false,
        "isOptional": false,
        "isPrivate": false,
        "isProtected": false,
        "isPublic": false,
        "isRest": false,
        "isStatic": false
      },
      "kind": "type alias",
      "name": "ChipOnRemoveEventT",
      "type": "RMWC.CustomEventT<>"
    },
    "ChipHTMLProps": {
      "fileName": "src/chip/index.tsx",
      "flags": {
        "isExported": true,
        "isExternal": false,
        "isOptional": false,
        "isPrivate": false,
        "isProtected": false,
        "isPublic": false,
        "isRest": false,
        "isStatic": false
      },
      "kind": "type alias",
      "name": "ChipHTMLProps",
      "type": "RMWC.HTMLProps<HTMLDivElement, Omit<AllHTMLAttributes<HTMLDivElement>, \"label\">>"
    }
  }
}
