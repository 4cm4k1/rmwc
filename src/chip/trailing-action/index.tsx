import React from 'react';
import { createComponent } from '@rmwc/base';
import { MDCChipTrailingActionFoundation } from '@material/chips';

/*********************************************************************
 * Trailing Action
 *********************************************************************/

// export interface TrailingActionProps {
//   isNavigable?: boolean;
// }

export interface TrailingActionApi {
  getFoundation: () => MDCChipTrailingActionFoundation;
}

export interface TrailingActionProps {
  apiRef?: (api: TrailingActionApi | null) => void;
  isNavigable?: boolean;
  icon: string;
}

/** The trailing action is used in removable input chips.
 * It is a subcomponent of the chips and intended only for use in the context of a chip. */
export const TrailingAction = createComponent<TrailingActionProps>(
  function TrailingAction(props) {
    return (
      <button
        className="mdc-chip-trailing-action"
        aria-hidden={props.isNavigable}
      >
        <span className="mdc-chip-trailing-action__ripple"></span>
        <span className="mdc-chip-trailing-action__icon material-icons">
          {props.icon}
        </span>
      </button>
    );
  }
);
