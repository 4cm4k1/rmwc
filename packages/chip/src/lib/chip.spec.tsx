import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import { Chip } from './chip';
import { ChipSet } from './chip-set';

describe('Chip', () => {
  it('renders', () => {
    const { asFragment } = render(
      <ChipSet>
        <Chip icon="favorite" trailingIcon="close" label="test-label" />
      </ChipSet>
    );

    expect(screen.getByText('test-label')).toBeInTheDocument();

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders with children', () => {
    render(
      <ChipSet>
        <Chip icon="favorite" trailingIcon="close">
          test-label
        </Chip>
      </ChipSet>
    );

    expect(screen.getByText('test-label')).toBeInTheDocument();
  });

  it('renders as filter chips', () => {
    const { asFragment } = render(
      <ChipSet filter>
        <Chip icon="favorite" trailingIcon="close">
          test-label
        </Chip>
      </ChipSet>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders as input chips', () => {
    const { asFragment } = render(
      <ChipSet input>
        <Chip icon="favorite" trailingIcon="close">
          test-label
        </Chip>
      </ChipSet>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('handles selected', () => {
    render(
      <ChipSet>
        <Chip checkmark selected label="test-label" />
        <Chip checkmark selected icon="favorite" label="test-label" />
      </ChipSet>
    );
    expect(screen.getAllByRole('row')[0]).toHaveClass(
      'mdc-evolution-chip--selected'
    );
    expect(screen.getAllByRole('row')[1]).toHaveClass(
      'mdc-evolution-chip--selected'
    );
  });

  it('handles onInteraction', async () => {
    const onInteraction = vi.fn();
    render(<Chip label="my label" onInteraction={onInteraction} />);

    userEvent.click(screen.getByText('my label'));

    await waitFor(() => expect(onInteraction).toHaveBeenCalledTimes(1));
  });

  it('handles onRemove', async () => {
    const onRemove = vi.fn();
    render(
      <ChipSet input>
        <Chip label="my label" onRemove={onRemove} trailingIcon="close" />
      </ChipSet>
    );

    userEvent.click(screen.getByText('close'));

    await waitFor(() => expect(onRemove).toHaveBeenCalledTimes(1));
  });

  it('handles onInteraction', async () => {
    const onInteraction = vi.fn();
    render(
      <ChipSet>
        <Chip label="my label" onInteraction={onInteraction} />
      </ChipSet>
    );

    userEvent.click(screen.getByText('my label'));

    await waitFor(() => expect(onInteraction).toHaveBeenCalledTimes(1));
  });

  it('handles custom ChipIcon', () => {
    render(<Chip icon="favorite" />);

    expect(screen.getByText('favorite')).toBeInTheDocument();
  });

  it('should not trigger events when disabled', () => {
    const onClick = vi.fn();

    render(<Chip disabled label="Cookie" onClick={onClick} />);

    userEvent.click(screen.getByText('Cookie'));

    expect(onClick).not.toHaveBeenCalled();
  });

  it('can be overlow', () => {
    const { container } = render(
      <ChipSet overflow>
        <Chip label="Cookie" />
      </ChipSet>
    );
    expect(container.firstChild).toHaveClass('mdc-evolution-chip-set--overlow');
  });

  it('can be role grid', () => {
    const { container } = render(
      <ChipSet action>
        <Chip label="Cookie" />
      </ChipSet>
    );
    expect(container.firstChild).toHaveAttribute('role', 'grid');
  });

  it('can be role listbox and receive prop aria-orientation', () => {
    const { container } = render(
      <ChipSet filter>
        <Chip label="Cookie" />
      </ChipSet>
    );
    expect(container.firstChild).toHaveAttribute('role', 'listbox');
    expect(container.firstChild).toHaveAttribute(
      'aria-orientation',
      'horizontal'
    );
  });
});
