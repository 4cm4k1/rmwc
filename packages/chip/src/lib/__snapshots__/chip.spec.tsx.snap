// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Chip > renders 1`] = `
<DocumentFragment>
  <span
    class="mdc-evolution-chip-set"
    role="grid"
  >
    <span
      class="mdc-evolution-chip-set__chips"
      role="presentation"
    >
      <span
        class="mdc-ripple-surface mdc-evolution-chip"
        id="chip-test-label"
        role="row"
      >
        <span
          class="mdc-evolution-chip__cell mdc-evolution-chip__cell--primary"
          role="gridcell"
        >
          <button
            class="mdc-evolution-chip__action mdc-evolution-chip__action--primary"
            icon="favorite"
            label="test-label"
            role="gridcell"
            type="button"
          >
            <span
              class="mdc-evolution-chip__ripple mdc-evolution-chip__ripple--primary"
            />
            <span
              class="mdc-evolution-chip__graphic"
            >
              <i
                class="rmwc-icon rmwc-icon--ligature material-icons mdc-evolution-chip__icon mdc-evolution-chip__icon--primary rmwc-chip--action-input"
              >
                favorite
              </i>
            </span>
            <span
              class="mdc-evolution-chip__text-label"
            >
              test-label
            </span>
          </button>
        </span>
      </span>
    </span>
  </span>
</DocumentFragment>
`;

exports[`Chip > renders as filter chips 1`] = `
<DocumentFragment>
  <span
    aria-orientation="horizontal"
    class="mdc-evolution-chip-set"
    role="listbox"
  >
    <span
      class="mdc-evolution-chip-set__chips"
      role="presentation"
    >
      <span
        class="mdc-ripple-surface mdc-evolution-chip mdc-evolution-chip--filter"
        id="chip-:r2:"
        role="presentation"
      >
        <span
          class="mdc-evolution-chip__cell mdc-evolution-chip__cell--primary"
          role="gridcell"
        >
          <button
            class="mdc-evolution-chip__action mdc-evolution-chip__action--primary"
            icon="favorite"
            role="option"
            type="button"
          >
            <span
              class="mdc-evolution-chip__ripple mdc-evolution-chip__ripple--primary"
            />
            <span
              class="mdc-evolution-chip__graphic"
            >
              <i
                class="rmwc-icon rmwc-icon--ligature material-icons mdc-evolution-chip__icon mdc-evolution-chip__icon--primary"
              >
                favorite
              </i>
            </span>
            <span
              class="mdc-evolution-chip__text-label"
            >
              test-label
            </span>
          </button>
        </span>
      </span>
    </span>
  </span>
</DocumentFragment>
`;

exports[`Chip > renders as input chips 1`] = `
<DocumentFragment>
  <span
    class="mdc-evolution-chip-set"
    role="grid"
  >
    <span
      class="mdc-evolution-chip-set__chips"
      role="presentation"
    >
      <span
        class="mdc-ripple-surface mdc-evolution-chip rmwc-chip"
        id="chip-:r3:"
        role="row"
      >
        <span
          class="mdc-evolution-chip__cell mdc-evolution-chip__cell--primary"
          role="gridcell"
        >
          <button
            class="mdc-evolution-chip__action mdc-evolution-chip__action--primary"
            icon="favorite"
            role="gridcell"
            type="button"
          >
            <span
              class="mdc-evolution-chip__ripple mdc-evolution-chip__ripple--primary"
            />
            <span
              class="mdc-evolution-chip__graphic"
            >
              <i
                class="rmwc-icon rmwc-icon--ligature material-icons mdc-evolution-chip__icon mdc-evolution-chip__icon--primary rmwc-chip--action-input"
              >
                favorite
              </i>
            </span>
            <span
              class="mdc-evolution-chip__text-label"
            >
              test-label
            </span>
          </button>
        </span>
        <span
          class="mdc-evolution-chip__cell mdc-evolution-chip__cell--trailing"
          role="gridcell"
        >
          <button
            class="mdc-evolution-chip__action mdc-evolution-chip__action--trailing"
            data-mdc-deletable="true"
            icon="close"
            type="button"
          >
            <span
              class="mdc-evolution-chip__ripple mdc-evolution-chip__ripple--trailing"
            />
            <i
              class="rmwc-icon rmwc-icon--ligature material-icons mdc-evolution-chip__icon mdc-evolution-chip__icon--trailing"
            >
              close
            </i>
          </button>
        </span>
      </span>
    </span>
  </span>
</DocumentFragment>
`;
